ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParsePayload.java

```diff
@@ -97,11 +104,29 @@ protected PubsubMessage processElement(PubsubMessage element)
     // Throws IOException if not a valid json object
     final JSONObject json = Json.readJSONObject(element.getPayload());
 
-    Map<String, String> attributes = element.getAttributeMap();
-    if (attributes != null && !attributes.containsKey("document_version")) {
-      // This element must be from the /submit/telemetry endpoint;
-      // we need to version from the payload.
-      attributes = new HashMap<>(attributes);
+    Map<String, String> attributes;
+    if (element.getAttributeMap() == null) {
+      attributes = new HashMap<>();
+    } else {
+      attributes = new HashMap<>(element.getAttributeMap());
+    }
+
+    // Remove any top-level "metadata" field if it exists, and attempt to parse it as a
+    // key-value map of strings, adding all entries as attributes.
+    Object untypedMetadata = json.remove("metadata");
+    if (untypedMetadata instanceof JSONObject) {
+      JSONObject metadata = (JSONObject) untypedMetadata;
+      for (String key : metadata.keySet()) {
+        Object value = metadata.get(key);
+        if (value != null) {
+          attributes.put(key, metadata.get(key).toString());
+        }
+      }
+    }
```

Comment:
Doing this here is definitely imperfect, since we won't have access to these new attributes for any earlier stages of the pipeline. Thus, we wouldn't be able to run the full Decoder based on a file in `text` format because ParseUri would throw errors about not having a uri attribute.

But we could build a future variant of the pipeline that calls this `ParsePayload` transform without the preceding steps that exist in Decoder.

In any case, this seems a bit safer than just straight-up dropping a "metadata" field if it exists.